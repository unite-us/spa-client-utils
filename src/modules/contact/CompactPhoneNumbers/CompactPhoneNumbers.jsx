import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Button, Icon } from '@unite-us/ui';
import phones from 'modules/utils/phones';

const MAX_NUMBERS_COLLAPSED = 3;

const Caret = (
  <Icon
    icon="IconChevronDown"
    style={{
      svg: {
        display: 'inline-block',
        verticalAlign: 'bottom',
      },
    }}
  />
);

const CompactPhoneNumbers = ({
  phone_numbers = [],
}) => {
  if (!phone_numbers.length) {
    return null;
  }

  const [expanded, setExpanded] = useState(false);

  const rootClasses = classNames('ui-contact-information__compact-phone', {
    expanded,
  });

  const phoneNumbers = phones
    .sortPhoneNumbers(phone_numbers, [
      'mobile',
      'home',
      'work',
      'fax',
      'unknown',
    ])
    .filter(number => Boolean(number.phone_number));

  return (
    <div className={rootClasses}>
      <div className="ui-provider-card__icon-text--phone">
        <Icon icon="IconPhone" />
      </div>

      {phoneNumbers.map((phoneNumber, index) => {
        const {
          extension = '',
          phone_type = '',
          phone_number = '',
          is_primary = false,
        } = phoneNumber;

        if ((!expanded && (index + 1) > MAX_NUMBERS_COLLAPSED) || !phone_number) {
          return null;
        }

        return (
          <div className="phone-number" key={`phone-${phone_number}`}>
            {phone_type && (
              <span
                className="phone-number__phone-type"
                data-test-element={`phone-numbers_type_${index}`}
              >
                {phone_type}:{' '}
              </span>
            )}

            <a
              href={phones.getTelLink(phoneNumber)}
              target="_blank"
              tabIndex={0}
              rel="noopener noreferrer"
              data-test-element={`phone-numbers_link_${index}`}
            >
              <span
                data-test-element={`phone-numbers_number_${index}`}
              >
                {`${phones.formatPhoneNumber(phone_number)}${extension ? ` ${extension}` : ''}`}
              </span>
            </a>

            {is_primary && (
              <span
                className="phone-number__is-primary"
                data-test-element={`phone-numbers_primary_${index}`}
              >
                {' '}Primary
              </span>
            )}
          </div>
        );
      })}
      {phoneNumbers.length > MAX_NUMBERS_COLLAPSED && (
        <div
          className="phone-number__toggle"
          data-test-element="phone-numbers_toggle"
        >
          <Button
            iconRight={Caret}
            label={`See ${!expanded ? 'more' : 'less'}`}
            link
            onClick={() => setExpanded(!expanded)}
          />
        </div>
      )}
    </div>
  );
};

CompactPhoneNumbers.propTypes = {
  phone_numbers: PropTypes.arrayOf(PropTypes.shape({
    phone_type: PropTypes.string,
    phone_number: PropTypes.string,
    country_code: PropTypes.string,
    is_primary: PropTypes.bool,
    extension: PropTypes.string,
  })),
};

CompactPhoneNumbers.defaultProps = {
  phone_numbers: [],
};

export default CompactPhoneNumbers;
