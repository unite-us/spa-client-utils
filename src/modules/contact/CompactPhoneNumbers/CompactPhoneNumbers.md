Simple example of the component rendering 2 phones.
```jsx
const phone_numbers = [
  {
    phone_number: '1232221233',
    extension: null,
    phone_type: 'fax',
  },
  {
    phone_number: '1232221233',
    extension: null,
    phone_type: 'work',
  },
  {
    phone_number: '9845665154',
    extension: null,
    phone_type: 'mobile',
    is_primary: true
  },
];

<CompactPhoneNumbers phone_numbers={phone_numbers} />
```

When rendering 4 or more phone numbers, 3 numbers are shown, with a toggle button to view the rest.
```jsx
const phone_numbers = [
  {
    phone_number: '1232221233',
    phone_type: 'mobile',
  },
  {
    phone_number: '2223123454',
    phone_type: 'work',
  },
  {
    phone_number: '9845665154',
    phone_type: 'home',
    is_primary: true
  },
  {
    phone_number: '7564533215',
    phone_type: 'fax',
  },
];

<CompactPhoneNumbers phone_numbers={phone_numbers} showToggle />
```
