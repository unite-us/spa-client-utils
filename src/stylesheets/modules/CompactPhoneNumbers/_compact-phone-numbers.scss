@import '~@unite-us/ui/src/stylesheets/variables';

.ui-contact-information__compact-phone {
  position: relative;
  margin-bottom: 20px;
  line-height: 20px;

  .ui-provider-card__icon-text--phone {
    position: absolute;
    top: 0;
    left: 0;
  }
  
  .program-details__icon--phone {
    width: 16px;
    height: 16px;
  }

  .ui-button--link {
    padding: 0;
    margin: 0 auto;
    font-size: 14px;
  }

  .ui-button__icon-right {
    transform: none;
    svg {
      transform: translateY(-2px);
    }
  }

  &.expanded {
    .ui-button__icon-right {
      svg {
        transform: rotate(180deg) translateY(-1px);
      }
    }
  }

  .ui-button__icon-right,
  .ui-provider-card__icon-text--phone {
    svg {
      fill: $united-blue-color;
    }
  }

  .phone-number {
    padding-left: 30px;
    margin-top: 5px;
    line-height: 20px;
    a {
      white-space: nowrap;
    }

    &:first-child {
      margin-top: 0;
    }
    
    &__phone-type {
      text-transform: capitalize;
    }

    &__is-primary {
      font-style: italic;
    }

    &__toggle {
      margin-top: 5px;
    }
  }
}