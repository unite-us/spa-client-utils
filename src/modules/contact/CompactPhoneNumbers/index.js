import CompactPhoneNumbers from './CompactPhoneNumbers';

export default CompactPhoneNumbers;
