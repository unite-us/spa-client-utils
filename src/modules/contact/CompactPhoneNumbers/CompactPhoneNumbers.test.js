import React from 'react';
import renderer from 'react-test-renderer';
import CompactPhoneNumbers from './CompactPhoneNumbers';

describe('CompactPhoneNumbers', () => {
  const props = {
    phone_numbers: [],
  };

  beforeEach(() => {
    props.phone_numbers = [
      {
        phone_number: '1232221233',
        extension: '1',
        phone_type: 'home',
      },
      {
        phone_number: '2223123454',
        extension: '1',
        phone_type: 'unknown',
      },
      {
        phone_type: 'work',
        phone_number: '1234567890',
      },
      {
        phone_type: 'fax',
        phone_number: '1231234567',
      },
    ];
  });

  it('renders phone numbers', () => {
    expect(renderer.create(<CompactPhoneNumbers {...props} />)).toMatchSnapshot();
  });

  it('does not render anything if the phone numbers list is empty', () => {
    expect(
      renderer.create(<CompactPhoneNumbers phone_numbers={[]} />).toJSON(),
    ).toBe(null);
  });

  it('renders max 3 phone numbers', () => {
    expect(
      renderer.create((
        <CompactPhoneNumbers
          phone_numbers={props.phone_numbers}
        />
      ))
        .root
        .findAllByProps({ className: 'phone-number' }),
    ).toHaveLength(3);
  });

  it('does not render numbers with falsy number properties', () => {
    props.phone_numbers[0].phone_number = '';
    props.phone_numbers.splice(2, 2);
    expect(
      renderer.create((
        <CompactPhoneNumbers
          phone_numbers={props.phone_numbers}
        />
      ))
        .root
        .findAllByProps({ className: 'phone-number' }),
    ).toHaveLength(1);
  });

  it('renders phone_type property', () => {
    const instance = renderer.create(<CompactPhoneNumbers {...props} />);
    expect(instance.root.findAllByProps({ className: 'phone-number__phone-type' })[0].children[0])
      .toEqual(props.phone_numbers[0].phone_type);
  });

  it('renders a toggle button when there are more than 3 numbers', () => {
    const instance = renderer.create(<CompactPhoneNumbers {...props} />);
    expect(instance.root.findAllByProps({ className: 'phone-number__toggle' }))
      .toHaveLength(1);
  });

  it('does not render a toggle button when there are 3 numbers or less', () => {
    props.phone_numbers.splice(0, 1);
    const instance = renderer.create(<CompactPhoneNumbers {...props} />);
    expect(instance.root.findAllByProps({ className: 'phone-number__toggle' }))
      .toHaveLength(0);
  });
});
