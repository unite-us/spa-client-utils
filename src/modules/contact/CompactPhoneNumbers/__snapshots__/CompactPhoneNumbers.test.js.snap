// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompactPhoneNumbers renders phone numbers 1`] = `
<div
  className="ui-contact-information__compact-phone"
>
  <div
    className="ui-provider-card__icon-text--phone"
  >
    <svg
      className="ui-icon"
      height="16px"
      id=""
      style={
        Object {
          "display": "inline-block",
          "fill": "",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 32 32"
      width="16px"
    >
      <path
        d="M2.126 1.188c1.278-1.281 5.141-1.495 5.17-0.817 0.029 0.677 3.479 8.191 3.509 8.869s-2.274 2.987-2.948 3.663c-0.668 0.669 4.508 6.533 4.606 6.644 0.11 0.099 5.96 5.287 6.628 4.617 0.674-0.676 2.979-2.985 3.653-2.955 0.676 0.029 8.173 3.487 8.849 3.518 0.676 0.029 0.463 3.9-0.815 5.182-1.093 1.095-6.752 5.071-16.592-2.029-1.079-0.622-2.788-2.049-5.38-4.646-0.001-0.001-0.003-0.003-0.005-0.004-0.002-0.003-0.004-0.004-0.006-0.006s-0.004-0.004-0.006-0.006c-0.001-0.002-0.003-0.004-0.004-0.005-2.59-2.598-4.014-4.312-4.635-5.393-7.083-9.864-3.116-15.537-2.024-16.632z"
      />
    </svg>
  </div>
  <div
    className="phone-number"
  >
    <span
      className="phone-number__phone-type"
      data-test-element="phone-numbers_type_0"
    >
      home
      :
       
    </span>
    <a
      data-test-element="phone-numbers_link_0"
      href="#"
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        data-test-element="phone-numbers_number_0"
      >
        (123) 222-1233 1
      </span>
    </a>
  </div>
  <div
    className="phone-number"
  >
    <span
      className="phone-number__phone-type"
      data-test-element="phone-numbers_type_1"
    >
      work
      :
       
    </span>
    <a
      data-test-element="phone-numbers_link_1"
      href="#"
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        data-test-element="phone-numbers_number_1"
      >
        (123) 456-7890
      </span>
    </a>
  </div>
  <div
    className="phone-number"
  >
    <span
      className="phone-number__phone-type"
      data-test-element="phone-numbers_type_2"
    >
      fax
      :
       
    </span>
    <a
      data-test-element="phone-numbers_link_2"
      href="#"
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        data-test-element="phone-numbers_number_2"
      >
        (123) 123-4567
      </span>
    </a>
  </div>
  <div
    className="phone-number__toggle"
    data-test-element="phone-numbers_toggle"
  >
    <button
      aria-label="See more"
      className="ui-button ui-button--icon ui-button--link"
      data-test-element={null}
      disabled={false}
      id=""
      onClick={[Function]}
      style={Object {}}
      type="button"
    >
      See more
      <span
        className="ui-button__icon-right"
      >
        <svg
          className="ui-icon"
          height="16px"
          id=""
          style={
            Object {
              "display": "inline-block",
              "verticalAlign": "bottom",
            }
          }
          viewBox="0 0 28 28"
          width="16px"
        >
          <path
            d="M26.297 12.625l-11.594 11.578c-0.391 0.391-1.016 0.391-1.406 0l-11.594-11.578c-0.391-0.391-0.391-1.031 0-1.422l2.594-2.578c0.391-0.391 1.016-0.391 1.406 0l8.297 8.297 8.297-8.297c0.391-0.391 1.016-0.391 1.406 0l2.594 2.578c0.391 0.391 0.391 1.031 0 1.422z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
